File: naive_cpp/compute.cpp
#include "compute.hpp"

#include <complex>
#include <algorithm>


int* compute(Resolution resolution, View view, int maxIterations) {
	int size = std::max(resolution.Width, resolution.Height);
	double scale = 3 / view.zoom;

	int* iterations = new int[resolution.Width * resolution.Height];

	int index = 0;
	for (int y = 0; y < resolution.Height; y++) {
		for (int x = 0; x < resolution.Width; x++) {
			std::complex<double> c = std::complex<double>(
				(x + .5 - resolution.Width / 2) / size * scale + view.center_x,
				(y + .5 - resolution.Height / 2) / size * scale + view.center_y
			);
			std::complex<double> z;

			int iteration;
			for (iteration = 0; iteration < maxIterations; iteration++) {
				z = z * z + c;
				if (z.real() * z.real() + z.imag() * z.imag() > 2) {
					break;
				}
			}

			iterations[index++] = iteration;
		}
	}

	return iterations;
}

File: naive_python/compute.py
def compute(resolution, view, max_iterations):
    width, height = resolution
    size = max(width, height)

    (center_x, center_y), zoom = view
    scale = 3 / zoom

    iterations = [[0] * height for _ in range(0, width)]
    for y in range(0, height):
        for x in range(0, width):
            c = complex((x + .5 - width / 2) / size * scale + center_x, (y + .5 - height / 2) / size * scale + center_y)
            z = complex(0, 0)

            for iteration in range(0, max_iterations):
                z = z * z + c
                if z.real * z.real + z.imag * z.imag > 2:
                    break

            iterations[x][y] = iteration

    return iterations
